File: sdk/go/stepauth/validate.go
package stepauth

import (
	"fmt"
	"net/url"
	"regexp"
	"time"
	"unicode/utf8"
)

// loopbackRe matches an http loopback host (localhost/127.0.0.1), the hub's
// dev allowance for callbackUrl.
var loopbackRe = regexp.MustCompile(`^(localhost|127\.0\.0\.1)$`)

// Validate runs the pre-send checks the hub itself enforces, so a caller
// gets a local error instead of a round trip. It returns error rather than
// *ProtocolError so a caller wrapping it in an error-returning function does
// not turn a valid request into a typed-nil failure; recover the code with
// errors.As.
func Validate(req *AuthorizationRequest) error {
	if req == nil {
		return NewProtocolError(CodeMalformedRequest, "request is nil")
	}
	if err := validateFields(req); err != nil {
		return err
	}
	if err := validateTiming(req, time.Now()); err != nil {
		return err
	}
	if err := validateSubject(req); err != nil {
		return err
	}
	if req.CallbackURL != "" {
		if err := validateCallbackURL(req.CallbackURL); err != nil {
			return err
		}
	}
	if !validEntries(req.Principal.Attributes) || !validEntries(req.Action.Details) {
		return NewProtocolError(CodeMalformedRequest, "labeled entry keys must be unique within their array")
	}
	return nil
}

// validateFields checks required-field presence, id charset, action.name
// length, and the closed category registry.
func validateFields(req *AuthorizationRequest) *ProtocolError {
	switch {
	case req.RequestID == "":
		return NewProtocolError(CodeMalformedRequest, "missing requestId")
	case req.SenderID == "":
		return NewProtocolError(CodeMalformedRequest, "missing senderId")
	case req.RecipientID == "":
		return NewProtocolError(CodeMalformedRequest, "missing recipientId")
	case req.Timestamp == "":
		return NewProtocolError(CodeMalformedRequest, "missing timestamp")
	case req.ExpiresAt == "":
		return NewProtocolError(CodeMalformedRequest, "missing expiresAt")
	case req.Action.Type == "":
		return NewProtocolError(CodeMalformedRequest, "missing action.type")
	case req.Action.Category == "":
		return NewProtocolError(CodeMalformedRequest, "missing action.category")
	case req.Action.Summary == "":
		return NewProtocolError(CodeMalformedRequest, "missing action.summary")
	case req.Principal.Subject == "":
		return NewProtocolError(CodeMalformedRequest, "missing principal.subject")
	case !validKeyID(req.RequestID):
		return NewProtocolError(CodeMalformedRequest, "requestId has invalid characters or is too long")
	case !validKeyID(req.SenderID):
		return NewProtocolError(CodeMalformedRequest, "senderId has invalid characters or is too long")
	case utf8.RuneCountInString(req.Action.Name) > MaxActionNameLen:
		return NewProtocolError(CodeMalformedRequest, fmt.Sprintf("action.name is longer than %d characters", MaxActionNameLen))
	}
	if !IsValidCategory(req.Action.Category) {
		return NewProtocolError(CodeUnknownCategory, "unknown action.category: "+req.Action.Category)
	}
	return nil
}

// validKeyID matches the hub's requestId/senderId charset: non-empty, at
// most MaxKeyIDLen bytes, printable ASCII, no '#' (a sort-key delimiter on
// the hub side that would let one id's key range straddle another's).
func validKeyID(id string) bool {
	if id == "" || len(id) > MaxKeyIDLen {
		return false
	}
	for i := range len(id) {
		if c := id[i]; c == '#' || c < 0x20 || c > 0x7e {
			return false
		}
	}
	return true
}

// validateTiming checks that timestamp and expiresAt parse, timestamp skew
// against now, then operator-chain depth.
func validateTiming(req *AuthorizationRequest, now time.Time) *ProtocolError {
	ts, err := time.Parse(time.RFC3339, req.Timestamp)
	if err != nil {
		return NewProtocolError(CodeTimestampOutOfRange, "timestamp is not RFC3339")
	}
	if d := now.Sub(ts); d > MaxClockSkew || d < -MaxClockSkew {
		return NewProtocolError(CodeTimestampOutOfRange, "timestamp skew exceeds 5m")
	}
	if _, err := time.Parse(time.RFC3339, req.ExpiresAt); err != nil {
		return NewProtocolError(CodeMalformedRequest, "expiresAt is not RFC3339")
	}

	depth := 0
	for p := req.Principal.Operator; p != nil; p = p.Operator {
		depth++
		if depth > MaxOperatorDepth {
			return NewProtocolError(CodeOperatorDepthExceeded, "operator chain deeper than 5")
		}
	}
	return nil
}

// validateSubject checks principal.subject is a valid individual NameID, and
// action.approver, if set, is a valid NameID.
func validateSubject(req *AuthorizationRequest) *ProtocolError {
	n, err := ParseNameID(req.Principal.Subject)
	if err != nil || !n.IsIndividual() {
		return NewProtocolError(CodeMalformedRequest, "principal.subject must be an individual NameID")
	}
	if req.Action.Approver != "" && !ValidNameID(req.Action.Approver) {
		return NewProtocolError(CodeMalformedRequest, "action.approver is not a valid NameID")
	}
	return nil
}

// validateCallbackURL enforces https, or http to a loopback host as a dev allowance.
func validateCallbackURL(raw string) *ProtocolError {
	u, err := url.Parse(raw)
	if err != nil {
		return NewProtocolError(CodeInvalidCallbackURL, "callbackUrl is not a valid URL")
	}
	if !httpsOrLoopback(u) {
		return NewProtocolError(CodeInvalidCallbackURL, "callbackUrl must be https, or http to localhost/127.0.0.1")
	}
	return nil
}

// httpsOrLoopback is the scheme rule shared by callbackUrl and hub host:
// https, or http to a loopback host as a dev allowance.
func httpsOrLoopback(u *url.URL) bool {
	if u.Scheme == "https" {
		return true
	}
	return u.Scheme == "http" && loopbackRe.MatchString(u.Hostname())
}

File: sdk/es/src/stepauth/validate.ts
import { MAX_CLOCK_SKEW_MS } from "./wire.js";
import type { AuthorizationRequest, LabeledEntry, Principal } from "./wire.js";
import { isValidCategory } from "./category.js";
import { isIndividual, isValidNameID, parseNameID } from "./nameid.js";
import { validateEntries } from "./entry.js";
import { ProtocolError } from "./errors.js";

const MAX_KEY_ID_LEN = 256;
/** action.name bound, in codepoints. */
const MAX_ACTION_NAME_LEN = 120;
const MAX_OPERATOR_DEPTH = 5;

const RFC3339_RE =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/;

function isRFC3339(s: string): boolean {
  return RFC3339_RE.test(s) && !Number.isNaN(Date.parse(s));
}

/** Codepoint count, matching Go's utf8.RuneCountInString: `s.length` counts
 * UTF-16 code units and over-counts anything outside the BMP (emoji, many CJK
 * extension characters), so TS and Go would disagree on identical input. */
function runeLength(s: string): number {
  return [...s].length;
}

/** ids spliced into DynamoDB sort keys: non-empty, bounded, printable ASCII,
 * no '#' (the sort-key delimiter — an embedded one could straddle another
 * id's key range). */
function isValidKeyID(id: string): boolean {
  if (id === "" || id.length > MAX_KEY_ID_LEN) return false;
  for (let i = 0; i < id.length; i++) {
    const c = id.charCodeAt(i);
    if (c === 0x23 || c < 0x20 || c > 0x7e) return false;
  }
  return true;
}

function malformed(message: string): never {
  throw new ProtocolError("malformed_request", message);
}

function callbackAllowed(url: URL): boolean {
  if (url.protocol === "https:") return true;
  return url.protocol === "http:" && (url.hostname === "localhost" || url.hostname === "127.0.0.1");
}

function validateCallbackURL(callbackUrl: string | undefined): void {
  if (!callbackUrl) return;
  let url: URL;
  try {
    url = new URL(callbackUrl);
  } catch {
    throw new ProtocolError("invalid_callback_url", "callbackUrl is not a valid URL");
  }
  if (!callbackAllowed(url)) {
    throw new ProtocolError("invalid_callback_url", "callbackUrl must be https");
  }
}

function validateOperatorDepth(principal: Principal): void {
  let depth = 0;
  for (let p = principal.operator; p; p = p.operator) {
    depth++;
    if (depth > MAX_OPERATOR_DEPTH) {
      throw new ProtocolError("operator_depth_exceeded", "operator chain deeper than 5");
    }
  }
}

function validateLabeledEntries(items: LabeledEntry[] | undefined): void {
  if (!items) return;
  try {
    validateEntries(items);
  } catch (err) {
    malformed(err instanceof Error ? err.message : "invalid labeled entry");
  }
}

/**
 * Pre-send validation mirroring the hub's own checks. Throws a ProtocolError
 * with the code the hub would return for the same defect.
 */
export function validateRequest(req: AuthorizationRequest): void {
  if (req.requestId === "") malformed("missing requestId");
  if (req.senderId === "") malformed("missing senderId");
  if (req.recipientId === "") malformed("missing recipientId");
  if (req.timestamp === "") malformed("missing timestamp");
  if (req.expiresAt === "") malformed("missing expiresAt");
  if (req.action.type === "") malformed("missing action.type");
  if ((req.action.category as string) === "") malformed("missing action.category");
  if (req.action.summary === "") malformed("missing action.summary");
  if (req.principal.subject === "") malformed("missing principal.subject");

  if (!isValidKeyID(req.requestId)) malformed("requestId has invalid characters or is too long");
  if (!isValidKeyID(req.senderId)) malformed("senderId has invalid characters or is too long");

  if (runeLength(req.action.name ?? "") > MAX_ACTION_NAME_LEN) {
    malformed(`action.name is longer than ${MAX_ACTION_NAME_LEN} characters`);
  }
  if (!isValidCategory(req.action.category)) {
    throw new ProtocolError("unknown_category", `unknown action.category: ${String(req.action.category)}`);
  }

  if (!isRFC3339(req.timestamp)) {
    throw new ProtocolError("timestamp_out_of_range", "timestamp is not RFC3339");
  }
  // Checked against the local clock. The hub's clock is authoritative; this
  // only catches a stale or wrongly built timestamp before it costs a round
  // trip.
  if (Math.abs(Date.now() - Date.parse(req.timestamp)) > MAX_CLOCK_SKEW_MS) {
    throw new ProtocolError("timestamp_out_of_range", "timestamp skew exceeds 5m");
  }
  if (!isRFC3339(req.expiresAt)) malformed("expiresAt is not RFC3339");

  validateOperatorDepth(req.principal);

  const subject = isValidNameID(req.principal.subject) ? parseNameID(req.principal.subject) : null;
  if (!subject || !isIndividual(subject)) {
    malformed("principal.subject must be an individual NameID");
  }
  if (req.action.approver && !isValidNameID(req.action.approver)) {
    malformed("action.approver must be a valid NameID");
  }

  validateCallbackURL(req.callbackUrl);

  validateLabeledEntries(req.principal.attributes);
  validateLabeledEntries(req.action.details);
}
